### PR TITLE
add convience methods to store for cancellation and loading state

### PR DESCRIPTION
**Pull request by buntec**
Experimental!

### CODE REVIEW

File: ff4s/src/main/scala/ff4s/CancellationKey.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ff4s
+
+import cats.Eq
+import cats.Show
+
+final case class CancellationKey(key: String)
```
Comment (armanbilge) [inline ff4s/src/main/scala/ff4s/CancellationKey.scala:22]: I wonder if instead of `String` it should be a `Unique.Token`?

Comment (buntec) [reply ff4s/src/main/scala/ff4s/CancellationKey.scala]: How about this?

File: ff4s/src/main/scala/ff4s/CancellationKey.scala [hunk 2]
```diff
@@ -16,14 +16,20 @@
...
+)
 
 object CancellationKey {
 
-  implicit val show: Show[CancellationKey] = Show.show(_.key)
+  /** Generates a unique cancellation key. */
+  def apply[F[_]: Unique: Applicative]: F[CancellationKey] =
+    Unique[F].unique.map(new CancellationKey(_))
 
   implicit val eq: Eq[CancellationKey] = Eq.fromUniversalEquals
```
Comment (armanbilge) [inline ff4s/src/main/scala/ff4s/CancellationKey.scala:34]:
```suggestion
  implicit val eq: Eq[CancellationKey] = Eq.by(_.token)
```

**Review by armanbilge — COMMENTED** [review]
Dumb question: how is this different from `Hotswap` with `fu.background`?

**Comment by buntec** [comment]
Not sure. I want the effect returned by `withCancellationKey` to run exactly as long as the effect being wrapped, and if the former is cancelled, it should forward the cancellation to the latter. (This allows wrapping one effect with two different cancellation keys, for example, and getting the correct loading state irrespective of the order of `withLoading` and `withCancellationKey`). With a `Supervisor` we can achieve this, I think, by doing `fiber.join.onCancel(fiber.cancel)`. The `fiber.join` bit should be possible with the hotswap, but not sure about forwarding cancellation 🤔 .

**Comment by buntec** [comment]
Many thanks to @ramytanios and @armanbilge for valuable feedback!

**Pull request merged**